File: src/com/douma/line/algo/linkedlist/_23_MergeKLinkedLists.java
package com.douma.line.algo.linkedlist;

/**
 * @微信公众号 : 抖码课堂
 * @官方微信号 : bigdatatang01
 * @作者 : 老汤
 */
public class _23_MergeKLinkedLists {
    // 分治思想
    // 时间复杂度：O(k*n*logk)
    // 空间复杂度：O(logk)
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0)
            return null;
        return merge(lists, 0, lists.length - 1);
    }

    private ListNode merge(ListNode[] lists, int left, int right) {
        if (left == right) return lists[left];

        int mid = left + (right - left) / 2;

        ListNode mergedLeftList = merge(lists, left, mid);
        ListNode mergedRightList = merge(lists, mid + 1, right);

        return mergeTwoLists(mergedLeftList, mergedRightList);
    }

    // 顺序合并
    // 时间复杂度：O(k^2 * n)
    // 空间复杂度：O(1)
    public ListNode mergeKLists1(ListNode[] lists) {
        if (lists == null || lists.length == 0)
            return null;
        ListNode outputList = lists[0];
        for (int i = 1; i < lists.length; i++) {
            outputList = mergeTwoLists(outputList, lists[i]);
        }
        return outputList;
    }

    // 时间复杂度：O(2n)
    // 空间复杂度：O(1)
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        if (l2 == null) return l1;

        ListNode dummyNode = new ListNode(-1);
        ListNode curr = dummyNode;

        while (l1 != null && l2 != null) {
            if (l1.val <= l2.val) {
                curr.next = l1;
                l1 = l1.next;
            } else {
                curr.next = l2;
                l2 = l2.next;
            }
            curr = curr.next;
        }

        if (l1 == null) curr.next = l2;
        if (l2 == null) curr.next = l1;

        return dummyNode.next;
    }
}

File: python_code/line/algo/linkedlist/_23_MergeKLinkedLists.py
import heapq
from typing import List

from line.algo.linkedlist.ListNode import ListNode


class Solution:

    # 分治思想
    # 时间复杂度：O(k*n*logk)
    # 空间复杂度：O(logk)
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if len(lists) == 0: return None
        return self.merge(lists, 0, len(lists) - 1)

    def merge(self, lists: List[ListNode], left: int, right: int) -> ListNode:
        if left == right: return lists[left]

        mid = left + (right - left) // 2
        merged_left_list = self.merge(lists, left, mid)
        merged_right_list = self.merge(lists, mid + 1, right)

        return self.mergeTwoLists(merged_left_list, merged_right_list)

    # 顺序合并
    # 时间复杂度：O(k^2 * n)
    # 空间复杂度：O(1)
    def mergeKLists1(self, lists):
        if len(lists) == 0: return None
        outputList = lists[0]
        for i in range(1, len(lists)):
            outputList = self.mergeTwoLists(outputList, lists[i])
        return outputList

    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        if not l1: return l2
        if not l2: return l1

        dummy_node = ListNode(-1)
        curr = dummy_node
        while l1 and l2:
            if l1.val <= l2.val:
                curr.next = l1
                l1 = l1.next
            else:
                curr.next = l2
                l2 = l2.next
            curr = curr.next

        if l1: curr.next = l1
        if l2: curr.next = l2

        return dummy_node.next
